Replace runtime strip_path function with compiler intrinsic equivalents

scartmell-arm opened the pull request:
### Description

#### Replace runtime strip_path function with compiler intrinsic equivalents

Sleep manager tracing strips the path from filenames and uses the result as an identifier to track drivers that unlock/lock sleep tracing. Replace the function that strips the path from the string, replace this function with a new macro, `__FILENAME__` which performs the same action in a compiler specific manner.

- GCC_ARM, use __builtin_strrchr which is optimized out at compile time.
- ARM, use `__MODULE__` which returns the filename without a path.
- IAR, specify the --no_path_in_file_macros compiler flag.

#### Change the driver identifier in the sleep tracing manager to a pointer of the stored file path.

Currently, the identifier is 15-byte character array that is copied from the `__FILE__` of the driver being profiled. Given that the content of `__FILE__` is stored inside the binary for the duration of the application it would be more efficient to store a pointer to this character string. This saves space in the statistics struct and saves time during lookup and insertion.


### Pull request type

- [ ] Fix
- [x] Refactor
- [ ] New target
- [ ] Feature
- [ ] Breaking change

@deepikabhavnani @SenRamakri @bulislaw

Commit:
Replace runtime strip_path function with compiler intrinsic equivalents



Sleep manager tracing strips the path from filenames and uses the result as an
identifier to track drivers that unlock/lock sleep tracing. Replace the function
that strips the path from the string, replace this function with a new macro,
__FILENAME__ which performs the same action in a compiler specific manner.

- GCC_ARM, use __builtin_strrchr which is optimized out at compile time.
- ARM, use __MODULE__ which returns the filename without path.
- IAR, specifiy the --no_path_in_file_macros compiler flag.

Commit:
Refactor sleep tracing driver identifier to be pointer to the driver filepath.

The use of __FILE__ macro to get a usable identifier from the driver path
causes the path of the file to be stored in the .text region of the binary.
Given that this remains for the entire duration of the program, storing a
pointer to this string as an identifier is more efficient than copying the
contents of the string during lookup/insertion.

Commit:
Remove superfluous compiler check in macro

deepikabhavnani reviewed (dismissed):
Looks good to me

0xc0170 reviewed (commented):
Looks fine, but the change `--no_path_in_file_macros` in  the compiler flags for IAR - is this required for this patch to land?

Changing the compiler flags sets the minor version label for this

Commit:
Remove IAR compiler flag change from PR

Commit:
Fix issues with __FILENAME__ macro

- Move macro definition to mbed_toolchain.h
- Remove double underscores from macro which are reserved.
- Fix macro for IAR until compiler flags to disable path are added again.
